explainer.md

```diff
@@ -182,6 +182,50 @@ The main benefit to this approach is that it would ensure that developers don't
 
 As an aside, the `notificationclick` event has similar challenges to the `launch` event in that handlers can be written such that nothing happens when a notification is clicked. Whatever solution is decided for `launch` event should also apply to `notificationclick` for consistency.
 
+### Should apps that aren't installed have launch events fired?
+
+It is unclear whether launch events should be gated on installation. Ultimately this should be up to the user agent, but it would be nice to be able to provide some kind of recommendation. 
+
+Our preference here is that apps which are not installed don't receive launch events, so as to break user expectations of the web as little as possible (installed apps are a new thing). In general, we prefer not to gate APIs on installation, however, there is some precedence for granting more permissions upon installation; PiP and WebShareTarget, for example.
+
+### Potentially Breaking `window.open`
+`window.open` is a synchronous  API, while the nature of launch events requires them to be asynchronous. It is unclear how we will be able to reconcile these two APIs (blocking until the launch event completes doesn't seem super acceptable). This could break popup based authentication flows (and potentially more).
```

Comment:
Not so much break `window.open` as break with `window.open`. The two APIs together result in some unexpected behavior.

Having to wait for redirects is a valid concern, which I think is going to require some discussion. 

That said, I don't think we should be waiting for a navigation before firing the launch event: The whole point is that we might not want to do the navigation (i.e. we parse the url and postMessage instead). 

This is how I imagine the flow working:

1. Site calls `window.open('http://google.com')`
2. Chrome checks for an installed ServiceWorker @ google.com
    1. Launch Event is fired --> We focus some new/existing client window, or show a notification
    2. or, if there is no ServiceWorker, we open a window and begin a navigation

Thinking about it like this, maybe we can return `null` from `window.open` if the navigation is handled by a `launch` event (if sites need communication between windows, they can postMessage)? What are your thoughts?

Reply:
> That said, I don't think we should be waiting for a navigation before firing the launch event: The whole point is that we might not want to do the navigation (i.e. we parse the url and postMessage instead).

Agreed - I didn't mean that we should always wait for the navigation to happen. I just meant that in same cases you /will/ have to wait for the navigation to start. In particular, if I have https://www.google.com/app as my installed PWA and a website calls window.open('https://google.com'), those are two different origins and we can't assume that the www.google.com app can handle google.com links. I have to wait until the redirect to www.google.com actually happens, at which point I can intercept the navigation and fire a launch event.

I agree that ideally you would be able to intercept the launch event before the window is opened. In cases where it is possible to do that (URL in window.open() matches) we might be able to do something. One fear is that window opening logic is extremely complicated (especially with site isolation). I think this makes me lean toward trying to handle windows (at least as a first cut) in the same way we would handle any other navigation. i.e. actually open the window and then intercept the navigation. This would be much simpler as a first cut.